plugin/secondary: Retries on i/o-timeout on startup

Is there a way to configure the `secondary`-plugin to do automatic retries if it fails to load a zone-file?
Or is there some way that I can trigger a reload on zones that failed?

I'm currently setting up a CoreDNS server with secondary, to handle primary DNS outages, but I sometimes get data like this:

```plain
CoreDNS-1.8.4
linux/amd64, go1.16.4, 053c4d5
[INFO] plugin/file: Transferred: <host1>. from <axfr-host>:53
[INFO] plugin/file: Transferred: <host2>. from <axfr-host>:53
[INFO] plugin/file: Transferred: <host3>. from <axfr-host>:53
[INFO] plugin/file: Transferred: <host4>. from <axfr-host>:53
[INFO] plugin/file: Transferred: <host5>. from <axfr-host>:53
[INFO] plugin/file: Transferred: <host6>. from <axfr-host>:53
[ERROR] plugin/file: Failed to transfer `<host7>.' from "<axfr-host>:53": read tcp <local-ip>:33580-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host8>.' from "<axfr-host>:53": read tcp <local-ip>:33602-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host9>.' from "<axfr-host>:53": read tcp <local-ip>:33596-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host10>.' from "<axfr-host>:53": read tcp <local-ip>:33598-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host11>.' from "<axfr-host>:53": read tcp <local-ip>:33590-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host12>.' from "<axfr-host>:53": read tcp <local-ip>:33610-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host13>.' from "<axfr-host>:53": read tcp <local-ip>:33586-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host14>.' from "<axfr-host>:53": read tcp <local-ip>:33608-><axfr-host>:53: i/o timeout
[ERROR] plugin/file: Failed to transfer `<host15>.' from "<axfr-host>:53": read tcp <local-ip>:33582-><axfr-host>:53: i/o timeout
[INFO] plugin/file: Transferred: <host16>. from <axfr-host>:53
```

The DNS provider is DNSimple.

--- a/plugin/secondary/setup.go
+++ b/plugin/secondary/setup.go
@@ -1,6 +1,7 @@
 package secondary
 
 import (
+	"k8s.io/apimachinery/pkg/util/wait"
 	"time"
 
 	"github.com/coredns/caddy"
@@ -29,13 +30,18 @@ func setup(c *caddy.Controller) error {
 			c.OnStartup(func() error {
 				z.StartupOnce.Do(func() {
 					go func() {
+						backoff := wait.Backoff{
+							Duration: time.Millisecond * 250,
+							Factor:   2,
+							Cap:      time.Second * 10,
+						}
 						for {
 							err := z.TransferIn()
 							if err == nil {
 								break
 							}
 							log.Warningf("All '%s' masters failed to transfer: %s, retrying in 10s", n, err)
-							time.Sleep(time.Second * 10)
+							time.Sleep(backoff.Step())
 						}
 						z.Update()
 					}()

--- a/plugin/secondary/setup.go
+++ b/plugin/secondary/setup.go
@@ -1,6 +1,8 @@
 package secondary
 
 import (
+	"time"
+
 	"github.com/coredns/caddy"
 	"github.com/coredns/coredns/core/dnsserver"
 	"github.com/coredns/coredns/plugin"
@@ -8,7 +10,6 @@ import (
 	clog "github.com/coredns/coredns/plugin/pkg/log"
 	"github.com/coredns/coredns/plugin/pkg/parse"
 	"github.com/coredns/coredns/plugin/pkg/upstream"
-	"time"
 )
 
 var log = clog.NewWithPlugin("secondary")

--- a/plugin/secondary/setup.go
+++ b/plugin/secondary/setup.go
@@ -10,8 +10,6 @@ import (
 	clog "github.com/coredns/coredns/plugin/pkg/log"
 	"github.com/coredns/coredns/plugin/pkg/parse"
 	"github.com/coredns/coredns/plugin/pkg/upstream"
-
-	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 var log = clog.NewWithPlugin("secondary")
@@ -31,18 +29,20 @@ func setup(c *caddy.Controller) error {
 			c.OnStartup(func() error {
 				z.StartupOnce.Do(func() {
 					go func() {
-						backoff := wait.Backoff{
-							Duration: time.Millisecond * 250,
-							Factor:   2,
-							Cap:      time.Second * 10,
-						}
+						dur := time.Millisecond * 250
+						step := time.Duration(2)
+						max := time.Second * 10
 						for {
 							err := z.TransferIn()
 							if err == nil {
 								break
 							}
-							log.Warningf("All '%s' masters failed to transfer: %s, retrying in 10s", n, err)
-							time.Sleep(backoff.Step())
+							log.Warningf("All '%s' masters failed to transfer, retrying in %s: %s", n, dur.String(), err)
+							time.Sleep(dur)
+							dur = step * dur
+							if dur > max {
+								dur = max
+							}
 						}
 						z.Update()
 					}()

--- a/plugin/secondary/setup.go
+++ b/plugin/secondary/setup.go
@@ -1,7 +1,6 @@
 package secondary
 
 import (
-	"k8s.io/apimachinery/pkg/util/wait"
 	"time"
 
 	"github.com/coredns/caddy"
@@ -11,6 +10,8 @@ import (
 	clog "github.com/coredns/coredns/plugin/pkg/log"
 	"github.com/coredns/coredns/plugin/pkg/parse"
 	"github.com/coredns/coredns/plugin/pkg/upstream"
+
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 var log = clog.NewWithPlugin("secondary")

--- a/plugin/secondary/setup.go
+++ b/plugin/secondary/setup.go
@@ -5,9 +5,13 @@ import (
 	"github.com/coredns/coredns/core/dnsserver"
 	"github.com/coredns/coredns/plugin"
 	"github.com/coredns/coredns/plugin/file"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
 	"github.com/coredns/coredns/plugin/pkg/parse"
 	"github.com/coredns/coredns/plugin/pkg/upstream"
+	"time"
 )
+
+var log = clog.NewWithPlugin("secondary")
 
 func init() { plugin.Register("secondary", setup) }
 
@@ -24,7 +28,14 @@ func setup(c *caddy.Controller) error {
 			c.OnStartup(func() error {
 				z.StartupOnce.Do(func() {
 					go func() {
-						z.TransferIn()
+						for {
+							err := z.TransferIn()
+							if err == nil {
+								break
+							}
+							log.Warningf("All '%s' masters failed to transfer: %s, retrying in 10s", n, err)
+							time.Sleep(time.Second * 10)
+						}
 						z.Update()
 					}()
 				})

--- a/test/secondary_test.go
+++ b/test/secondary_test.go
@@ -132,3 +132,87 @@ func TestIxfrResponse(t *testing.T) {
 		t.Fatalf("Serial should be %d, got %d", 2015082541, soa.Serial)
 	}
 }
+
+func TestRetryInitialTransfer(t *testing.T) {
+	// Start up a secondary that expects to transfer from a master that doesn't exist yet
+	corefile := `example.org:0 {
+		secondary {
+			transfer from 127.0.0.1:5399
+		}
+	}`
+
+	i, udp, _, err := CoreDNSServerAndPorts(corefile)
+	if err != nil {
+		t.Fatalf("Could not get CoreDNS serving instance: %s", err)
+	}
+	defer i.Stop()
+
+	m := new(dns.Msg)
+	m.SetQuestion("www.example.org.", dns.TypeA)
+	resp, err := dns.Exchange(m, udp)
+	if err != nil {
+		t.Fatal("Expected to receive reply, but didn't")
+	}
+	// Expect that the query will fail
+	if resp.Rcode != dns.RcodeServerFailure {
+		t.Fatalf("Expected reply to be a SERVFAIL, got %d", resp.Rcode)
+	}
+
+	// Now spin up the master server
+	name, rm, err := test.TempFile(".", `$ORIGIN example.org.
+@ 3600 IN SOA  sns.dns.icann.org. noc.dns.icann.org. (
+        2017042745 ; serial
+        7200       ; refresh (2 hours)
+        3600       ; retry (1 hour)
+        1209600    ; expire (2 weeks)
+        3600       ; minimum (1 hour)
+)
+
+  3600 IN NS   a.iana-servers.net.
+  3600 IN NS   b.iana-servers.net.
+
+www    IN A    127.0.0.1
+www    IN AAAA ::1
+`)
+	if err != nil {
+		t.Fatalf("Failed to create zone: %s", err)
+	}
+	defer rm()
+
+	corefileMaster := `example.org:5399 {
+		file ` + name + `
+		transfer {
+          to *
+        }
+	}`
+
+	master, _, _, err := CoreDNSServerAndPorts(corefileMaster)
+	if err != nil {
+		t.Fatalf("Could not start CoreDNS master: %s", err)
+	}
+	defer master.Stop()
+
+	retry := time.Tick(time.Millisecond * 100)
+	timeout := time.Tick(time.Second * 5)
+
+	for {
+		select {
+		case <-retry:
+			m = new(dns.Msg)
+			m.SetQuestion("www.example.org.", dns.TypeA)
+			resp, err = dns.Exchange(m, udp)
+			if err != nil {
+				continue
+			}
+			// Expect the query to succeed
+			if resp.Rcode != dns.RcodeSuccess {
+				continue
+			}
+			return
+		case <-timeout:
+			t.Fatal("Timed out trying for successful response.")
+			return
+		}
+	}
+
+}

